GPMSA Example: Brian Williams' linear verification

briadam opened the pull request:
@roystgnr @dmcdougall @brianw525 This starts to address #579 by implementing Brian's test problem in a QUESO example.  It runs and doesn't crash (issues a bunch of warnings from one particular spot though.)

In terms of your development workflow, I wasn't sure if you want this to be managed on a branch until the test is verified vs. merged to devel. Feel free to guide me as to how to work.

Add initial version of Bayes linear verification problem due to Brian
Williams

[1] Adams, B.M., Coleman, Kayla, Hooper, Russell W., Khuwaileh, Bassam
A., Lewis, Allison, Smith, Ralph C., Swiler, Laura P., Turinsky, Paul
J., Williams, Brian J., "User Guidelines and Best Practices for CASL
VUQ Analysis Using Dakota, Sandia Technical Report 2016-11614 (also
CASL-U-2016-1233-000), September 2016.

briadam committed:
GPMSA Example: Brian Williams' linear verification

Add initial version of Bayes linear verification problem due to Brian
Williams

[1] Adams, B.M., Coleman, Kayla, Hooper, Russell W., Khuwaileh, Bassam
A., Lewis, Allison, Smith, Ralph C., Swiler, Laura P., Turinsky, Paul
J., Williams, Brian J., "User Guidelines and Best Practices for CASL
VUQ Analysis Using Dakota, Sandia Technical Report 2016-11614 (also
CASL-U-2016-1233-000), September 2016.

briadam commented:
Spoke with @roystgnr who suggested if possible we make some or all of this a test, with XFAIL status if needed.  While the ultimate test @brianw525 has in mind likely won't be feasible for this (requires 100k or 250k samples) and is why I targeted examples/, we can make a scaled down version to use as at least a regression test.

dmcdougall commented:
Do you want me to merge this example or wait for a test?

briadam commented:
Yes, but I defer to you all as to what's appropriate in the project.  One of my goals prior to our next in-person meeting is to implement a few of Brian Williams' verification cases so we can actually iterate on working code in the meeting and run cases.  Since the cases have to run many samples, perhaps for now they belong as examples and not tests and that's just okay?  Also, since I'll be adding a few more cases similar to this, I am keenly interested in feedback on my implementation before I copy/paste and make other cases, but that's not required.

dmcdougall commented on examples/gp/linear_verif/gpmsa_input.txt:
```diff
@@ -0,0 +1,69 @@
+###############################################
+#ip_mh_help                 = anything
+ip_mh_dataOutputFileName   = outputData/sipOutput
+ip_mh_dataOutputAllowedSet = 0 1
```
Vectors need single quotes around them (`'0 1'`) because`dev` uses `GetPot` to parse input files, and not `boost`.  You can still use `boost` if you want, but the default is `GetPot`.

dmcdougall commented on examples/gp/linear_verif/gpmsa_input.txt:
```diff
@@ -0,0 +1,69 @@
+ip_mh_rawChain_displayPeriod        = 2000
+ip_mh_rawChain_measureRunTimes      = 1
+ip_mh_rawChain_dataOutputFileName   = outputData/ip_raw_chain
+ip_mh_rawChain_dataOutputAllowedSet = 0 1
```
Needs quotes.

dmcdougall commented on examples/gp/linear_verif/gpmsa_input.txt:
```diff
@@ -0,0 +1,69 @@
+ip_mh_tk_useLocalHessian            = 0
+ip_mh_tk_useNewtonComponent         = 0
+ip_mh_dr_maxNumExtraStages          = 0
+ip_mh_dr_listOfScalesForExtraStages = 5. 10. 20.
```
Needs quotes.

dmcdougall commented on examples/gp/linear_verif/gpmsa_linear_verif.C:
```diff
@@ -0,0 +1,345 @@
+// Bayes linear verification problem due to Brian Williams
```
Can you put the QUESO licence before this?  You can copy it from one of the other examples.

dmcdougall reviewed (changes requested):
You can add your example to [this file](https://github.com/libqueso/queso/blob/dev/examples/run_all.sh) if you want it to be easily executed from our `run_all.sh` script.

Caveat other remarks, this looks fine to me.  Thanks!

briadam commented:
Thanks for the helpful feedback.  Brian Williams updated this example and gave me several others, so I'm thinking I'll also reorganize the directories and do some cleanup, maybe even build some utilities that can be shared among the cases, before we merge...

dmcdougall commented:
No worries.  Let me know when you're ready to pull the trigger.

briadam committed:
GPMSA linear verification (scalar): address feedback

Address feedback from @dmcdougall in libqueso/queso#580
 * Add license
 * Quote vectors in text input file

briadam committed:
GPMSA linear verification: refactor for multi-response

Refactor linear verification data read to work for both scalar and
multivariate cases (number of responses > 1).  Rename input and data
files in preparation for adding multivariate case.

briadam committed:
GPMSA linear verification: setup driver for scalar vs MV

Minor rework of main to drive scalar vs. multivariate cases.

briadam committed:
GPMSA linear verification: add multivariate case

Add multivariate five-response linear verification case (example
toggles based on input filename for now), including observation error
covariance matrix.  Case is almost same as scalar, but just copied the
code for now to make debugging more straightforward.

Also add comments with @brianw525 recommended changes to priors and
initial point (tried changing them, but then scalar case doesn't run)

briadam committed:
GPMSA verification: address review @dmcdougall comments

 * Use @brianw525 recommended prior and initial point settings.
 * Scale experiment observation error covariance by simulation variance.

roystgnr commented on examples/Makefile.am:
```diff
@@ -444,6 +444,21 @@ dist_gpmsa_vector_DATA += gp/scalar/dakota_pstudy.dat
+gpmsa_linverif_CPPFLAGS = -I$(top_srcdir)/examples/gp/linear_verif $(AM_CPPFLAGS)
+
+dist_gpmsa_linverif_DATA =
+dist_gpmsa_linverif_DATA += ${gpmsa_scalar_SOURCES}
```
This installs gpmsa_scalar.C, not gpmsa_linear_verif.C

briadam committed:
GPMSA verification: fix install rule, loop ordering

Minor edits per reviewer feedback
